### PR TITLE
Auto Enable All Instrumentation Feature

### PR DESCRIPTION
**Pull request by ericmustin**
### Summary 

This PR adds functionality to automatically enable all available integrations without needing to add a `Datadog.configure` block.

This works by adding a script that can be `require`'d in an application startup lifecyce, which adds a `railtie` which iterates over all available integrations (except for those blacklisted), and enabled them automatically, silencing any logs that might be normally emitted from  a failed patching attempt.

In non rails environments, the feature does not use a railtie, and is still functional but requires users to `require` their integration's gems first.

The docs are also updated to include usage instructions.

### Notes

- There's still some tests here I'd like to add, But i was hoping to have the bulk of this work reviewed first for sanity. the current tests pass fine on it's own, but there appears to be leakage of config between rails tests. If you notice, you can intentionally remove the `require 'ddtrace/auto_instrument'` line from the `support/rails[3|4|5|6]` specs and the tests will still pass. Isolating the test via `fit` makes them pass/fail as expected, so I believe the rails application initialisation is getting leaked between spec tests. I tried wrapping the auto_instrument tests in a forked process but that didn't appear to help. Still investigating this

- cleaned up some odd behavior in our `Rake` integration where we were not being defensive enough

### CODE REVIEW

File: docs/GettingStarted.md [hunk 1]
```diff
@@ -132,6 +132,28 @@ Before downloading tracing on your application, install the Datadog Agent. The R
...
+    gem 'ddtrace'
+    ```
+
+2. Install the gem with `bundle install`
+
+3. Add `require 'ddtrace/auto_instrument'` to your application startup sequence. This can be done in your `config/application.rb` file.
+
+    ```ruby
+      require 'rails/all'
+      require 'ddtrace/patch_all'
```
Comment (brettlangdon) [inline docs/GettingStarted.md:150]:
```suggestion
      require 'ddtrace/auto_instrument'
```

Comment (brettlangdon) [reply docs/GettingStarted.md:150]: Python profiling has something similar, https://ddtrace.readthedocs.io/en/stable/basic_usage.html#via-module

File: docs/GettingStarted.md [hunk 2]
```diff
@@ -153,6 +175,20 @@ Before downloading tracing on your application, install the Datadog Agent. The R
...
 ### Quickstart for Ruby applications
 
+#### Ruby Auto Instrument all Integrations
+
+1. Install the gem with `gem install ddtrace`
+2. Add `require 'ddtrace/auto_instrument'` to your application startup sequence. This must be done _after_ requiring any libraries or frameworks that should be instrumented. 
+
+    ```ruby
+      require 'sinatra'
+      require 'ddtrace/patch_all'
```
Comment (brettlangdon) [inline docs/GettingStarted.md:185]:
```suggestion
      require 'ddtrace/auto_instrument'
```

**Comment by marcotc** [comment]
@ericmustin if you need to `require 'ddtrace/auto_instrument'`, but then "unrequire" it in a subsequent test, you best option is to isolate the tests that permanently pollute the workspace in their own set of Rake tasks.

We do that for `bundle exec appraisal rails6-postgres rake spec:railsdisableenv` for example, because we need a completely clean execution (no Rails ever initalized) for these tests.

File: lib/ddtrace/auto_instrument.rb [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+      Datadog.configure do |c|
+        c.silence_patching_logs
+        # This will activate auto-instrumentation for Rails
+        integrations.each do |integration_name|
+          c.use integration_name
+        end
+      end
+    end
+
+    if defined?(Rails) && defined?(Rails::VERSION) && defined?(Rails::VERSION::MAJOR) &&
```
Comment (marcotc) [inline lib/ddtrace/auto_instrument.rb:27]:
```suggestion
    if defined?(Rails::VERSION::MAJOR) &&
```
`defined?` is safe to use for nested name look-ups, even if the first parts of the constant don't exist.

File: lib/ddtrace/contrib/extensions.rb [hunk 4]
```diff
@@ -27,8 +27,9 @@ def configure(target = configuration, opts = {})
 
           # Activate integrations
           if target.respond_to?(:integrations_pending_activation)
+            silence_logs = target.respond_to?(:silence_logs?) ? target.silence_logs? : false
             target.integrations_pending_activation.each do |integration|
-              integration.patch if integration.respond_to?(:patch)
+              integration.patch(silence_logs) if integration.respond_to?(:patch)
```
Comment (marcotc) [inline lib/ddtrace/contrib/extensions.rb:32]: This is very tricky, as silencing all logs will also hide important information. For example, the instrumentation might not be compatible with the gem version installed (we support >= 1.0, but 0.9 is installed) (`Patchable.compatible?`), or the gem might not have been `require`d yet (`Patchable.loaded?`).

Both of these cases currently output a warning today.

Maybe we could have `Patchable#patch` return `true` is patching was successful, or a Hash if something failed:
```ruby
{
  available: self.class.available?,
  loaded: self.class.loaded?,
  compatible: self.class.compatible?,
  patchable: self.class.patchable?,
}
```
And let the caller handle the error reporting.

What do you think?

File: lib/ddtrace/contrib/extensions.rb [hunk 5]
```diff
@@ -86,6 +87,14 @@ def fetch_integration(name)
             registry[name] ||
               raise(InvalidIntegrationError, "'#{name}' is not a valid integration.")
           end
+
+          def silence_logs?
```
Comment (marcotc) [inline lib/ddtrace/contrib/extensions.rb:91]: I think `verbose?` seems like more straightforward name for this option (defaulting to true, and set to false when patching-all).

File: lib/ddtrace/contrib/extensions.rb (hunk 4 above)
Comment (ericmustin) [reply lib/ddtrace/contrib/extensions.rb]: makes sense, updated. I still think the logging is a little clunky but it's more flexible than previously.

File: lib/ddtrace/auto_instrument.rb (hunk 3 above)
Comment (ericmustin) [reply lib/ddtrace/auto_instrument.rb]: 👍 updated

File: docs/GettingStarted.md (hunk 2 above)
Comment (ericmustin) [reply docs/GettingStarted.md]: 👍

File: lib/ddtrace/contrib/extensions.rb [hunk 6]
```diff
@@ -27,8 +27,24 @@ def configure(target = configuration, opts = {})
 
           # Activate integrations
           if target.respond_to?(:integrations_pending_activation)
+            reduce_verbosity = target.respond_to?(:reduce_verbosity?) ? target.reduce_verbosity? : false
             target.integrations_pending_activation.each do |integration|
-              integration.patch if integration.respond_to?(:patch)
+              next unless integration.respond_to?(:patch)
+              # integration.patch returns either true or a hash of details on why patching failed
+              patch_results = integration.patch
+              next unless patch_results.is_a?(Hash)
```
Comment (marcotc) [inline lib/ddtrace/contrib/extensions.rb:35]:
```suggestion
              next if patch_results == true
```

File: spec/ddtrace/contrib/rails/rails_auto_instrument_spec.rb [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+  around do |example|
+    # Reset before and after each example; don't allow global state to linger.
+    Datadog.configuration.reset!
+    example.run
+    Datadog.configuration.reset!
+  end
+
+  context 'when auto patching is included' do
+    before do
+      skip 'Java not supported' if RUBY_PLATFORM == 'java'
```
Comment (marcotc) [inline spec/ddtrace/contrib/rails/rails_auto_instrument_spec.rb:15]:
```suggestion
      skip if PlatformHelpers.jruby?
```

I normally remove the `skip` message if the `if` condition is an exact copy of the conditional statement, but that's just a preference.

File: spec/ddtrace/contrib/rails/rails_auto_instrument_spec.rb [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+      allow(ENV).to receive(:[]).and_call_original
+      allow(ENV).to receive(:[]).with('TEST_AUTO_INSTRUMENT').and_return(true)
```
Comment (marcotc) [inline spec/ddtrace/contrib/rails/rails_auto_instrument_spec.rb:17]: `ClimateControl` allows us to mock environment variables, without having to worry if the exercises code is calling `ENV[]`, `ENV.fetch`, or `ENV.each`.
https://github.com/DataDog/dd-trace-rb/blob/f9530e3b1ab807e0e13e5f6f26c06e48b4f46f61/spec/ddtrace/contrib/rails/disable_env_spec.rb#L21-L23

File: docs/GettingStarted.md [hunk 9]
```diff
@@ -132,6 +132,28 @@ Before downloading tracing on your application, install the Datadog Agent. The R
...
+    gem 'ddtrace'
+    ```
+
+2. Install the gem with `bundle install`
+
+3. Add `require 'ddtrace/auto_instrument'` to your application startup sequence. This can be done in your `config/application.rb` file.
+
+    ```ruby
+      require 'rails/all'
+      require 'ddtrace/auto_instrument'
```
Comment (marcotc) [inline docs/GettingStarted.md:150]: Remove indentation level, as these require statements will be pasted at the top-level of their project file.
```suggestion
    require 'rails/all'
    require 'ddtrace/auto_instrument'
```

File: spec/ddtrace/contrib/rails/rails_auto_instrument_spec.rb (hunk 8 above)
Comment (ericmustin) [reply spec/ddtrace/contrib/rails/rails_auto_instrument_spec.rb]: makes sense, updated

File: spec/ddtrace/contrib/rails/rails_auto_instrument_spec.rb (hunk 7 above)
Comment (ericmustin) [reply spec/ddtrace/contrib/rails/rails_auto_instrument_spec.rb]: makes sense, updated

File: docs/GettingStarted.md [hunk 10]
```diff
@@ -132,6 +132,28 @@ Before downloading tracing on your application, install the Datadog Agent. The R
 
 ### Quickstart for Rails applications
 
+#### Rails Auto Instrument all Integrations
+
+1. Add the `ddtrace` gem to your Gemfile:
+
+    ```ruby
+    source 'https://rubygems.org'
+    gem 'ddtrace'
```
Comment (marcotc) [inline docs/GettingStarted.md:141]: Does this pattern work with Rails?
```suggestion
    gem 'ddtrace', require: 'ddtrace/auto_instrument'
```
This will technically import `auto_instrument` when needed. I'm not sure if the import order would work correctly, but I think it's something worth exploring.

When trying this change, make sure to remove the changes added to `config/application.rb`, as they hopefully won't be needed.

File: lib/ddtrace/auto_instrument.rb [hunk 11]
```diff
@@ -0,0 +1,46 @@
+require 'ddtrace'
+
+module Datadog
+  # AutoInstrumentation enables all integrations
+  module AutoInstrument
+    EXCLUDED_RAILS_INTEGRATIONS = [:rack, :action_cable, :active_support, :action_pack, :action_view, :active_record].freeze
+    EXCLUDED_TEST_INTEGRATIONS = [:cucumber, :rspec].freeze
```
Comment (marcotc) [inline lib/ddtrace/auto_instrument.rb:7]: If we add a new Rails integration or new test integration, and forget to add them here, I don't think any tests would fail today.

I don't have a strong preference on what approach to take, but I think we should ensure that, when a new integration is added, either by us or community, there should be no surprises with `AutoInstrument` with the integration addition.

I can see this being done in different ways:
* Having an automatic way to skip Rails sub-integrations and skip test integrations.
* Having integrations declare that they are allowed to be automatically activated (probably through a new field in `Datadog::Contrib::Patchable::ClassMethods` or similar).
* Having a whitelist of instrumentations allowed to be automatically enabled. This one is a bit worse, as it required the knowledge of a separate file than the `contrib` files added in an integration PR, but still a valid solution.

Like I said, I have no strong preference here, but I think we should address this concern.

Comment (ericmustin) [reply lib/ddtrace/auto_instrument.rb:7]: yea, makes sense. I tend to think `Having integrations declare that they are allowed to be automatically activated (probably through a new field in Datadog::Contrib::Patchable::ClassMethods or similar).` is the right approach here, as the `rails + test libraries` non-auto-instrumentation logic is pretty arbitrary, and i'd rather this just be declared on a per integration basis explicitly. i'll add this

File: docs/GettingStarted.md (hunk 10 above)
Comment (ericmustin) [reply docs/GettingStarted.md:141]: i'll test this, this makes life easier i think

File: spec/ddtrace/auto_instrument_spec.rb [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+require 'ddtrace/auto_instrument'
+
+RSpec.describe 'Sinatra instrumentation with ActiveRecord' do
+  include Rack::Test::Methods
+
+  let(:options) { {} }
+
+  before do
+  end
+
```
Comment (marcotc) [inline spec/ddtrace/auto_instrument_spec.rb:22]:
```suggestion
```

File: docs/GettingStarted.md [hunk 13]
```diff
@@ -153,6 +175,20 @@ Before downloading tracing on your application, install the Datadog Agent. The R
 
 ### Quickstart for Ruby applications
 
+#### Ruby Auto Instrument all Integrations
+
+1. Install the gem with `gem install ddtrace`
+2. Add `require 'ddtrace/auto_instrument'` to your application startup sequence. This must be done _after_ requiring any libraries or frameworks that should be instrumented. 
+
+    ```ruby
+    require 'sinatra'
```
Comment (marcotc) [inline docs/GettingStarted.md:184]: If I understand correctly, sinatra is just an example here. We should add a comment about it here, so that folks know they don't need sinatra specifically.

File: docs/GettingStarted.md (hunk 10 above)
Comment (ericmustin) [reply docs/GettingStarted.md]: 👍 yup this works, updated

File: lib/ddtrace/auto_instrument.rb (hunk 11 above)
Comment (ericmustin) [reply lib/ddtrace/auto_instrument.rb]: I added a `auto_instrument?` method to `Patchable::InstanceMethods` that accomplishes this. It's a bit tricky to define the logic for the rails sub-modules (or those auto-enabled by rails), since they behave differently depending on whether rails/railtie is present. But i've added some helpers here to get this to work and think it' much cleaner now

File: spec/ddtrace/auto_instrument_spec.rb (hunk 12 above)
Comment (ericmustin) [reply spec/ddtrace/auto_instrument_spec.rb]: 👍 updated

File: docs/GettingStarted.md (hunk 13 above)
Comment (ericmustin) [reply docs/GettingStarted.md:177]: Yup, makes sense, i've added a note here explaining this, i think it's clearer now.

File: lib/ddtrace/auto_instrument.rb (hunk 11 above)
Comment (ericmustin) [reply lib/ddtrace/auto_instrument.rb]: one tradeoff here is that this PR is now quite a bit bigger but mostly it's adding boilerplate/repetitive methods to integration / integration_spec files. for your own sanity:

- looking at the `auto_instrument?` related changes to `integration/integration_spec.rb` files for a "rails-y" integration (like, active_record) is the same as all other "rails-y" integrations,  
- `cucumber / rspec` are changes identical,
- and then all the other integrations inherit the behavior and have the same base test added to their integration_spec.rb

File: lib/ddtrace/utils/rails.rb [hunk 14]
```diff
@@ -0,0 +1,12 @@
+module Datadog
+  module Utils
+    # Common rails-related utility functions.
+    module Rails
+      module_function
+
+      def railtie_supported?
```
Comment (marcotc) [inline lib/ddtrace/utils/rails.rb:7]: Because this file relates to an integration (Rails), it should be moved to the `contrib` directory. I noticed we already have a Rails `Util` module, so I suggest moving this method there: https://github.com/DataDog/dd-trace-rb/blob/v0.43.0/lib/ddtrace/contrib/rails/utils.rb

File: spec/ddtrace/contrib/active_record/integration_spec.rb [hunk 15]
```diff
@@ -56,6 +56,26 @@
...
+    end
+
+    context 'when within a rails application' do
+      before do
+        allow(Datadog::Utils::Rails).to receive(:railtie_supported?).and_return(true)
+      end
+
+      it { is_expected.to be(false) }
+    end
+  end
```
Comment (marcotc) [inline spec/ddtrace/contrib/active_record/integration_spec.rb:77]: I suggested you extract this into a shared example, as it seems like we use it with the exact same code in many Rails sub-gems.

File: docs/GettingStarted.md [hunk 16]
```diff
@@ -132,6 +132,21 @@ Before downloading tracing on your application, install the Datadog Agent. The R
 
 ### Quickstart for Rails applications
 
+#### Rails Auto Instrument all Integrations
+
+1. Add the `ddtrace` gem to your Gemfile:
+
+    ```ruby
+    source 'https://rubygems.org'
+    gem 'ddtrace', require: 'ddtrace/auto_instrument'
```
Comment (marcotc) [inline docs/GettingStarted.md:141]: This looks awesome! So much simpler than today 🎉.

**Review by marcotc — COMMENTED** [review]
I left 2 super `nit` comments. Everything else looks ready to ship 👍

File: spec/ddtrace/contrib/active_record/integration_spec.rb (hunk 15 above)
Comment (ericmustin) [reply spec/ddtrace/contrib/active_record/integration_spec.rb]: makes sense, updated

File: lib/ddtrace/utils/rails.rb (hunk 14 above)
Comment (ericmustin) [reply lib/ddtrace/utils/rails.rb]: yup agreed, updated

**Comment by ericmustin** [comment]
Active Model Serializer tests are all failing for 2.5/2.6/2.7 consistently with the same msg

```
      Failure/Error: attributes :name
      
      NoMethodError:
        undefined method `attributes' for #<Class:0x000055d8faca4760>
        Did you mean?  attribute_method?
```

My initial suspicion is that it looks like they cut a minor release this morning

https://rubygems.org/gems/active_model_serializers/versions/0.10.11

But looking for docs on it still or what changed in the api

**Comment by ericmustin** [comment]
The AMS stuff is a bit of a mess, looks like it's not super maintained at this point and they're working on 6.1 Rails support (which came out yday).  https://github.com/rails-api/active_model_serializers/pull/2398

Have to work through it still but have some mtgs, will followup later today

**Comment by ericmustin** [comment]
**Tl;dr** the rails 6.1 release may start to cause some transient failures in integration tests that depend on rails submodules, due to dependancy resolution issues.

The AMS issues (and maybe some other gems soon?) are because their latest version(`0.10.11`) has a dependancy on `activemodel` (a rails submodule) of `<6.1` in their latest release, but we're using rails submodules (activemodel, activesupport etc etc) of `6.1` in our appraisal env. So the AMS gem that's used is a very old one (`0.10.0`) which just didn't have any max version set. Unfort this means the internal interface we're expecting and we mock in our tests is different than the one we get, which causes errors.

AMS has already bumped support to `<6.2` but hasn't cut that release yet, I've pinged about this here and am waiting to hear back: https://github.com/rails-api/active_model_serializers/pull/2398#issuecomment-742613734

File: lib/ddtrace/auto_instrument.rb [hunk 17]
```diff
@@ -0,0 +1,41 @@
...
+      Datadog.configure do |c|
+        c.reduce_log_verbosity
+        # This will activate auto-instrumentation for Rails
+        integrations.each do |integration_name|
+          c.use integration_name
+        end
+      end
+    end
+
+    if Datadog::Contrib::Rails::Utils.railtie_supported?
```
Comment (delner) [inline lib/ddtrace/auto_instrument.rb:28]: I don't think this is a good idea. There's a deliberate design decision to keep integrations (Rails) separate from the core (`AutoInstrument`), and this creates coupling between the two. This means we always have to load Rails to do any kind of Ruby instrumentation, something we want to avoid if we want to move contributions eventually to their own separate package.

I would suggest instead implementing some kind of callback pattern: have the auto-instrumentation code here emit an  event in which the loaded Rails integration can hook into, receive, and act accordingly. It may not be totally that simple, but its definitely worthwhile avoiding this coupling.

**Review by delner — CHANGES_REQUESTED** [review]
The idea here is good and workable, but I don't think the implementation is right.

There's a deliberate design decision to keep integrations (Rails) separate from the core (`AutoInstrument`), and this creates coupling between the two. This means we always have to load Rails to do any kind of Ruby instrumentation, something we want to avoid if we want to move contributions eventually to their own separate package.

I would suggest instead implementing some kind of callback pattern: have the auto-instrumentation code emit an event in which the loaded Rails integration can hook into, receive, and act accordingly. Perhaps something else, as long as it avoids coupling the core to integration. It may not be totally that simple to do, but its definitely worthwhile avoiding this coupling.

File: lib/ddtrace/auto_instrument.rb [hunk 18]
```diff
@@ -0,0 +1,41 @@
...
+      # Railtie to include AutoInstrumentation in rails loading
+      class Railtie < Rails::Railtie
+        # we want to load before config initializers so that any user supplied config
+        # in config/initializers/datadog.rb will take precedence
+        initializer 'datadog.start_tracer', before: :load_config_initializers do
+          AutoInstrument.patch_all
+        end
+      end
```
Comment (marcotc) [inline lib/ddtrace/auto_instrument.rb:36]: _Following up on @delner's suggestion, about separating core from contrib._

We have 3 parts to this file which I think we can break apart:

1. This Rails-specific snippet. Can be extracted to the Rails sub-directory under contrib.
1. The `patch_all` implementation. This parts knows how to patch all instrumentations, but does not trigger itself. I don't know 100% where this would go, but `Datadog::Contrib::Extensions` looks very similar to where it could land. `Datadog::Contrib::Extensions` might be too overloaded for it, but I think it give us a good idea to where "integration core" features live today.
    To avoid coupling today `Datadog::Contrib::Extensions` injects itself into the global settings object: `Datadog::Configuration::Settings.send(:include, Configuration::Settings)`. This is a valid technique to prevent coupling.
1. A "product" part, where we activate it all. This is the code that is ultimately called when we require `ddtrace/auto_instrument`.

To make 1 and 2 work, one possible suggestion is to introduce another functionality to integrations: similarly to how they support `#patch` today, they could support something like `auto_patch`, and we could call that on all integrations that support it. This way new integration can choose to provide custom auto instrumentation or not.

File: lib/ddtrace/contrib/auto_instrument.rb [hunk 19]
```diff
@@ -0,0 +1,45 @@
...
+      # Patch adds method for invoking auto_instrumentation
+      module Patch
+        def add_auto_instrument
+          if Datadog::Contrib::Rails::Utils.railtie_supported?
+            require 'ddtrace/contrib/rails/auto_instrument_railtie'
+          else
+            AutoInstrument.patch_all
+          end
+        end
+      end
```
Comment (marcotc) [inline lib/ddtrace/contrib/auto_instrument.rb:21]: I think that the way we are composing `add_auto_instrument` doesn't seem like it would work nicely if another implementation tries to add new behaviour that also happen during `add_auto_instrument`.

I suggest we call `super` (if defined) in `add_auto_instrument` to ensure that the parent implementation is also called.

File: lib/ddtrace/auto_instrument.rb [hunk 20]
```diff
@@ -0,0 +1,3 @@
+require 'ddtrace'
+
+Datadog.add_auto_instrument if Datadog.respond_to?(:add_auto_instrument)
```
Comment (marcotc) [inline lib/ddtrace/auto_instrument.rb:3]: `Datadog.add_auto_instrument` is always defined now, as of latest changes, right?

File: lib/ddtrace/auto_instrument_base.rb [hunk 21]
```diff
@@ -0,0 +1,15 @@
+module Datadog
+  # base methods stubbed for adding auto instrument extensions
+  module AutoInstrumentBase
+
+    def self.included(base)
+      base.send(:extend, InstanceMethods)
+      base.send(:include, InstanceMethods)
+    end
```
Comment (marcotc) [inline lib/ddtrace/auto_instrument_base.rb:8]: Any particular reason to indirectly extend `InstanceMethods`, instead of defining `AutoInstrumentBase` as:
```ruby
module Datadog
  # base methods stubbed for adding auto instrument extensions
  module AutoInstrumentBase
    # stubbed methods for adding auto instrument
    def add_auto_instrument; end
  end
end
```

and then `extend`ing it on the base module:
```ruby
module Datadog
  extend AutoInstrumentBase
```

File: lib/ddtrace/auto_instrument.rb (hunk 20 above)
Comment (ericmustin) [reply lib/ddtrace/auto_instrument.rb]: yup, can remove the defensive check here now

File: lib/ddtrace/auto_instrument_base.rb (hunk 21 above)
Comment (ericmustin) [reply lib/ddtrace/auto_instrument_base.rb:7]: no not really, i was following a pattern i'd seen with some of the config modules, but this is far simpler so can update

File: lib/ddtrace/contrib/rails/auto_instrument_railtie.rb [hunk 22]
```diff
@@ -0,0 +1,10 @@
+require 'ddtrace'
+
+# Railtie to include AutoInstrumentation in rails loading
+class DatadogAutoInstrumentRailtie < Rails::Railtie
+  # we want to load before config initializers so that any user supplied config
+  # in config/initializers/datadog.rb will take precedence
+  initializer 'datadog.start_tracer', before: :load_config_initializers do
```
Comment (ericmustin) [inline lib/ddtrace/contrib/rails/auto_instrument_railtie.rb:7]: Just noting that a Railtie is necessary, as otherwise load order is done by rails alphabetically and we will not patch any gems loaded after ddtrace. this has been tested without a railtie and patching was unsuccesful on gems like faraday, http (anything alphabetically after dd-trace)

**Review by marcotc — APPROVED** [review]
🎉 thank you for working on the feedback provided, looks great!

File: docs/GettingStarted.md [hunk 23]
```diff
@@ -165,6 +180,25 @@ Install and configure the Datadog Agent to receive traces from your now instrume
 
 ### Quickstart for Ruby applications
 
+#### Ruby Auto Instrument all Integrations
```
Comment (andrewsouthard1) [inline docs/GettingStarted.md:183]: Should we not say here "Auto Instrument All Integrations". That's a bit wordy and inconsistent with how we describe in other languages. Maybe just "Automatic Instrumentation" and "Manual Instrumentation"

Comment (marcotc) [reply docs/GettingStarted.md:183]: It reads much better with your suggestion, probably an oversight on our side here. I'll make these changes. 🙇  @andrewsouthard1
